store/sqlstore/channel_store.go

```diff
@@ -3100,3 +3377,242 @@ func (s SqlChannelStore) GroupSyncedChannelCount() (int64, *model.AppError) {
+	var categories []*sidebarCategoryForJoin
+	if _, err := s.GetReplica().Select(&categories, "SELECT *, SidebarChannels.ChannelId FROM SidebarCategories, SidebarChannels WHERE SidebarCategories.UserId=:userId AND SidebarCategories.TeamId=:teamId AND SidebarChannels.CategoryId=SidebarCategories.Id AND SidebarCategories.Id=:categoryId ORDER BY SidebarChannels.SortOrder ASC", map[string]interface{}{"userId": userId, "teamId": teamId, "categoryId": categoryId}); err != nil {
+		return nil, model.NewAppError("SqlPostStore.GetSidebarCategories", "store.sql_channel.sidebar_categories.app_error", nil, err.Error(), http.StatusInternalServerError)
+	}
+	result := &model.SidebarCategoryWithChannels{
+		SidebarCategory: categories[0].SidebarCategory,
+		Channels:        make([]string, 0),
+	}
+	for _, category := range categories {
+		result.Channels = append(result.Channels, category.ChannelId)
+	}
+	return result, nil
+
+}
+
+func (s SqlChannelStore) GetSidebarCategories(userId, teamId string) (*model.OrderedSidebarCategories, *model.AppError) {
+	oc := model.OrderedSidebarCategories{
+		Categories: make(model.SidebarCategoriesWithChannels, 0),
+		Order:      make([]string, 0),
+	}
+
+	var categories []*sidebarCategoryForJoin
+	if _, err := s.GetReplica().Select(&categories, "SELECT *, SidebarChannels.ChannelId FROM SidebarCategories, SidebarChannels WHERE SidebarCategories.UserId=:userId AND SidebarCategories.TeamId=:teamId AND SidebarChannels.CategoryId=Id ORDER BY SidebarCategories.SortOrder ASC", map[string]interface{}{"userId": userId, "teamId": teamId}); err != nil {
+		return nil, model.NewAppError("SqlPostStore.GetSidebarCategories", "store.sql_channel.sidebar_categories.app_error", nil, err.Error(), http.StatusInternalServerError)
+	}
+	for _, category := range categories {
+		var prevCategory *model.SidebarCategoryWithChannels
+		for _, existing := range oc.Categories {
+			if existing.Id == category.Id {
+				prevCategory = existing
+				break
+			}
+		}
+		if prevCategory == nil {
+			prevCategory = &model.SidebarCategoryWithChannels{
+				SidebarCategory: category.SidebarCategory,
+				Channels:        make([]string, 0),
+			}
+			oc.Categories = append(oc.Categories, prevCategory)
+			oc.Order = append(oc.Order, category.Id)
+		}
+		prevCategory.Channels = append(prevCategory.Channels, category.ChannelId)
+	}
+	return &oc, nil
+}
+
+func (s SqlChannelStore) GetSidebarCategoryOrder(userId, teamId string) ([]string, *model.AppError) {
+	var ids []string
+	if _, err := s.GetReplica().Select(&ids, "SELECT Id FROM SidebarCategories WHERE UserId=:userId AND TeamId=:teamId ORDER BY SortOrder ASC", map[string]interface{}{"userId": userId, "teamId": teamId}); err != nil {
+		return nil, model.NewAppError("SqlPostStore.GetSidebarCategoryOrder", "store.sql_channel.sidebar_categories.app_error", nil, err.Error(), http.StatusInternalServerError)
```

Comment:
Do we want to check for a not found error and return a 404 ? A bad userid or teamId should not return a 500.